File: backend/src/main/java/com/backend/api/comment/service/CommentService.java
package com.backend.api.comment.service;

import com.backend.api.comment.dto.response.CommentMypageResponse;
import com.backend.api.comment.dto.response.CommentPageResponse;
import com.backend.api.comment.dto.response.CommentResponse;
import com.backend.api.post.service.PostService;
import com.backend.api.user.service.UserService;
import com.backend.domain.comment.entity.Comment;
import com.backend.domain.comment.repository.CommentRepository;
import com.backend.domain.post.entity.Post;
import com.backend.domain.user.entity.Role;
import com.backend.domain.user.entity.User;
import com.backend.global.Rq.Rq;
import com.backend.global.exception.ErrorCode;
import com.backend.global.exception.ErrorException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class CommentService {

    private final CommentRepository commentRepository;
    private final PostService postService;
    private final UserService userService;
    private final Rq rq;

    public Comment findByIdOrThrow(Long id) {
        return commentRepository.findById(id)
                .orElseThrow(() -> new ErrorException(ErrorCode.COMMENT_NOT_FOUND));
    }

    @Transactional
    public CommentResponse writeComment(User currentUser, Long postId, String content) {

        if (!currentUser.validateActiveStatus()) {
            throw new ErrorException(ErrorCode.ACCOUNT_SUSPENDED);
        }

        Post post = postService.findPostByIdOrThrow(postId);

        Comment comment = Comment.builder()
                .content(content)
                .author(currentUser)
                .post(post)
                .build();

        Comment savedComment = commentRepository.save(comment);

        return new CommentResponse(savedComment, null);
    }

    @Transactional
    public CommentResponse updateComment(User currentUser, Long commentId, String newContent) {
        Comment comment = this.findByIdOrThrow(commentId);

        if (!comment.getAuthor().getId().equals(currentUser.getId())) {
            throw new ErrorException(ErrorCode.COMMENT_INVALID_USER);
        }

        comment.updateContent(newContent);

        return new CommentResponse(comment, null);
    }

    @Transactional
    public void deleteComment(User currentUser, Long commentId) {
        Comment comment = this.findByIdOrThrow(commentId);

        if (!comment.getAuthor().getId().equals(currentUser.getId())) {
            throw new ErrorException(ErrorCode.COMMENT_INVALID_USER);
        }

        commentRepository.delete(comment);
    }

    public CommentPageResponse<CommentResponse> getCommentsByPostId(User currentUser, int page, Long postId) {
        postService.findPostByIdOrThrow(postId);

        if (page < 1) page = 1;
        Pageable pageable = PageRequest.of(page - 1, 20, Sort.by("createDate").ascending());

        Page<Comment> commentsPage = commentRepository.findByPostId(postId, pageable);

        Long currentUserId = currentUser != null ? currentUser.getId() : null;

        List<CommentResponse> comments = commentsPage.getContent()
                .stream()
                .map(comment -> {
                    boolean isMine = comment.getAuthor().getId().equals(currentUserId);
                    return new CommentResponse(comment, isMine);
                })
                .toList();

        return new CommentPageResponse<>(commentsPage, comments);
    }

    public CommentPageResponse<CommentMypageResponse> getCommentsByUserId(int page, Long userId) {
        userService.getUser(userId);
        User currentUser = rq.getUser();
        if(!currentUser.getId().equals(userId) && !currentUser.getRole().equals(Role.ADMIN)) {
            throw new ErrorException(ErrorCode.COMMENT_INVALID_USER);
        }

        if (page < 1) page = 1;
        Pageable pageable = PageRequest.of(page - 1, 15, Sort.by("createDate").descending());
        Page<Comment> userCommentsPage = commentRepository.findByAuthorId(userId, pageable);

        List<CommentMypageResponse> userComments = userCommentsPage.getContent()
                .stream()
                .map(CommentMypageResponse::new)
                .toList();

        return new CommentPageResponse<>(userCommentsPage, userComments);
    }
}

File: backend/src/main/java/com/backend/api/comment/service/CommentService.kt
package com.backend.api.comment.service

import com.backend.api.comment.dto.response.CommentMypageResponse
import com.backend.api.comment.dto.response.CommentPageResponse
import com.backend.api.comment.dto.response.CommentResponse
import com.backend.api.post.service.PostService
import com.backend.api.user.service.UserService
import com.backend.domain.comment.entity.Comment
import com.backend.domain.comment.repository.CommentRepository
import com.backend.domain.user.entity.Role
import com.backend.domain.user.entity.User
import com.backend.global.Rq.Rq
import com.backend.global.exception.ErrorCode
import com.backend.global.exception.ErrorException
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class CommentService(
    private val commentRepository: CommentRepository,
    private val postService: PostService,
    private val userService: UserService,
    private val rq: Rq
) {
    fun findByIdOrThrow(id: Long): Comment {
        return commentRepository.findById(id)
            .orElseThrow { ErrorException(ErrorCode.COMMENT_NOT_FOUND) }!!
    }

    @Transactional
    fun writeComment(currentUser: User, postId: Long, content: String): CommentResponse {
        if (!currentUser.validateActiveStatus()) {
            throw ErrorException(ErrorCode.ACCOUNT_SUSPENDED)
        }

        val post = postService.findPostByIdOrThrow(postId)

        val comment = Comment(
            content,
            post,
            currentUser,
        )

        val savedComment = commentRepository.save(comment)

        return CommentResponse.from(savedComment, null)
    }

    @Transactional
    fun updateComment(currentUser: User, commentId: Long, newContent: String): CommentResponse {
        val comment = this.findByIdOrThrow(commentId)

        if (comment.author.id  != currentUser.id) {
            throw ErrorException(ErrorCode.COMMENT_INVALID_USER)
        }

        comment.updateContent(newContent)

        return CommentResponse.from(comment, null)
    }

    @Transactional
    fun deleteComment(currentUser: User, commentId: Long) {
        val comment = this.findByIdOrThrow(commentId)

        if (comment.author.id  != currentUser.id) {
            throw ErrorException(ErrorCode.COMMENT_INVALID_USER)
        }

        commentRepository.delete(comment)
    }

    fun getCommentsByPostId(currentUser: User?, page: Int, postId: Long): CommentPageResponse<CommentResponse> {
        var page = page
        postService.findPostByIdOrThrow(postId)

        if (page < 1) page = 1
        val pageable: Pageable = PageRequest.of(page - 1, 20, Sort.by("createDate").ascending())

        val commentsPage: Page<Comment> = commentRepository.findByPostId(postId, pageable)

        val currentUserId: Long? = currentUser?.id

        val comments = commentsPage.getContent()
            .map  { comment: Comment ->
                val isMine = comment.author.id == currentUserId
                CommentResponse.from(comment, isMine)
            }

        return CommentPageResponse.from(commentsPage, comments)
    }

    fun getCommentsByUserId(page: Int, userId: Long): CommentPageResponse<CommentMypageResponse> {
        var page = page
        userService.getUser(userId)
        val currentUser = rq.getUser()
        if (currentUser.id != userId && currentUser.role != Role.ADMIN) {
            throw ErrorException(ErrorCode.COMMENT_INVALID_USER)
        }

        if (page < 1) page = 1
        val pageable: Pageable = PageRequest.of(page - 1, 15, Sort.by("createDate").descending())
        val userCommentsPage: Page<Comment> = commentRepository.findByAuthorId(userId, pageable)

        val userComments = userCommentsPage.getContent()
            .map { comment: Comment -> CommentMypageResponse.from(comment) }

        return CommentPageResponse.from(userCommentsPage, userComments)
    }
}